src/cosmic_ray/execution/local.py

```diff
@@ -93,25 +99,53 @@ def _execute_work_item(work_item):
     return work_item.job_id, result
 
 
+def _execute_no_mutate():
+    log.info('Executing worker in %s, PID=%s', _workspace.clone_dir, os.getpid())
+
+    with excursion(_workspace.clone_dir):
+        test_outcome, output = run_tests(_config.test_command, _config.timeout)
+        if test_outcome == TestOutcome.SURVIVED:
+            worker_outcome = WorkerOutcome.NORMAL
+        else:
+            worker_outcome = WorkerOutcome.ABNORMAL
+
+        return WorkResult(
+            output=output,
+            test_outcome=test_outcome,
+            worker_outcome=worker_outcome)
+
+
 class LocalExecutionEngine(ExecutionEngine):
     "The local-git execution engine."
 
     def __call__(self, pending_work, config, on_task_complete):
-        with multiprocessing.Pool(
+        asyncio.run(self._execute_pending_works(pending_work, config, on_task_complete))
+
+    async def _execute_pending_works(self,
+                                     pending_work: Iterable[WorkItem],
+                                     config: ConfigDict,
+                                     on_task_complete: Callable):
+        loop = asyncio.get_running_loop()
+
+        with concurrent.futures.ProcessPoolExecutor(
                 initializer=_initialize_worker,
-                initargs=(config,)) as pool:
-
-            # pylint: disable=W0511
-            # TODO: This is not optimal. The pending-work iterable could be millions
-            # or billions of elements. We don't want to copy it. We copy it right
-            # now so that we don't access the database in a separate thread (i.e.
-            # one created by imap_unoredered below). We need to find a way around
-            # this.
-            pending = list(pending_work)
-
-            results = pool.imap_unordered(
-                func=_execute_work_item,
-                iterable=pending)
-
-            for job_id, result in results:
-                on_task_complete(job_id, result)
+                initargs=(config,)
+        ) as pool:
+
+            if config.run_with_no_mutation:
+                log.info("Running initial work")
+                result = await loop.run_in_executor(pool, _execute_no_mutate)
+                log.info("Initial work ends with %s", result.worker_outcome)
+                if result.worker_outcome != WorkerOutcome.NORMAL:
+                    print("ERROR OCCURS DURING RUN WITH NO MUTATION")
+                    for line in result.output.split('\n'):
+                        print(" >>", line)
+                    return
+
+            async def run_task(work_item):
+                result = await loop.run_in_executor(pool, _execute_work_item,
+                                                    work_item)
+                on_task_complete(*result)
+
+            tasks = [run_task(work_item) for work_item in pending_work]
+            await asyncio.gather(*tasks)
```

Comment:
Why did you switch from multiprocessing to asyncio here? Was there something broken with the original version, or is this just personal preference? I'm not necessarily opposed to the switch, but I would absolutely prefer that you make these kinds of independent changes in separate PRs. As it is, you just make review more confusing.